File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:

        slow = head
        fast = head.next
        

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next


        temp = slow.next
        slow.next = None
        slow = temp

        if not slow:
            return



        # cut the end
        

        # flip the 2nd half
        prev = slow
        curr = prev.next
        prev.next = None

        while curr:
            temp = curr.next
            curr.next = prev

            prev = curr
            curr = temp

        right = prev
        left = head

        i = 0
        while left and right:

            if i % 2 == 0:
                temp = left.next
                left.next = right
                left = temp
            
            else:
                temp = right.next
                right.next = left
                right = temp

            i += 1

        return

        




            
        

File: Data Structures & Algorithms/add-two-numbers/submission-1.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} l1
     * @param {ListNode} l2
     * @return {ListNode}
     */
    addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode {

        let place = l1
        let sum = 0;
        let over = 0
        let prev;

        while (l1 && l2) {
            sum = l1.val + l2.val + over

            over = Math.trunc(sum / 10)
            sum = sum % 10

            l1.val = sum

            if (!l1.next) {
                prev = l1
            }

            l1 = l1.next
            l2 = l2.next
        }
        
        if (l2) {
            prev.next = l2
            l1 = l2
        }

        while (l1) {
            sum = l1.val + over
            over = Math.trunc(sum / 10)
            sum = sum % 10

            l1.val = sum

            if (!l1.next) {
                prev = l1
            }

            l1 = l1.next
        }

        if (over != 0) {
            const overNode = new ListNode(over)
            prev.next = overNode
        }

        return place;
    }
}
